### ISSUE TITLE
[Bug]: Custom logits processor edge-cases

### ISSUE THREAD

**Issue by afeldman-nm** [vllm-project/vllm#29045]
### Your current environment

N/A

### 🐛 Describe the bug

The following Logits Processor API bug was reported in vLLM Slack; reproducing this bug is TODO:

There are two edge cases where I see incorrect generations for my requests:

1. partial prefills: scheduler puts only a chunk of the prompt in the persistent batch. Model runner processes this, and then the sampler + LP runs, and the "wrong" token then gets discarded in the `_bookkeeping_sync` part of `gpu_model_runner.py`. The issue with this is that the partial prompt has modified state in the LP (`update_state()` was called on it), so even though the generated token was discarded the LP state is still corrupted.

2. preemptions: similarly, when a preempted request gets removed from the persistent batch, its LP state also gets deleted. When we re-add the preempted request later, it has a fresh state but the generated output tokens got added to the prompt (ie. we're not starting from scratch), so we essentially need to "replay" the generation inside the LP to restore state to before preemption.

For (2), I can think of some ways to add more bookkeeping inside my LP (for example, don't delete the request's state until I truly know the generation is finished), but I'm not sure what to do about (1). The `BatchUpdate` data structure gives us `prompt_tok_ids` and `output_tok_ids` for added requests, but from what I can tell: partial prefills and full prefills are indistinguishable in this scenario (they both have the same `prompt_tok_ids` and empty `output_tok_ids`). So I can't add code inside my LP to "skip state updates if we're processing a partial prefill". Do you have any suggestions for how I can deal with these scheduler edge cases? Let me know if theres any obvious fix I'm missing here as I'm relatively new to this part of the code

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.
